avgratecounter.go

```diff
@@ -2,39 +2,46 @@ package ratecounter
 
 import (
 	"strconv"
-	"sync/atomic"
 	"time"
 )
 
 // An AvgRateCounter is a thread-safe counter which returns
 // the ratio between the number of calls 'Incr' and the counter value in the last interval
 type AvgRateCounter struct {
-	hits     int64
-	counter  Counter
+	hits     *RateCounter
+	counter  *RateCounter
 	interval time.Duration
 }
 
 // NewRateCounter Constructs a new AvgRateCounter, for the interval provided
 func NewAvgRateCounter(intrvl time.Duration) *AvgRateCounter {
 	return &AvgRateCounter{
+		hits:     NewRateCounter(intrvl),
+		counter:  NewRateCounter(intrvl),
 		interval: intrvl,
 	}
 }
 
+func (r *AvgRateCounter) WithResolution(resolution int) *AvgRateCounter {
+	if resolution < 1 {
+		panic("AvgRateCounter resolution cannot be less than 1")
+	}
+
+	r.hits = r.hits.WithResolution(resolution)
+	r.counter = r.counter.WithResolution(resolution)
+
+	return r
+}
+
 // Incr Adds an event into the AvgRateCounter
 func (a *AvgRateCounter) Incr(val int64) {
-	atomic.AddInt64(&a.hits, 1)
+	a.hits.Incr(1)
 	a.counter.Incr(val)
-
-	time.AfterFunc(a.interval, func() {
-		atomic.AddInt64(&a.hits, -1)
-		a.counter.Incr(-1 * val)
-	})
 }
 
 // Rate Returns the current ratio between the events count and its values during the last interval
 func (a *AvgRateCounter) Rate() float64 {
-	hits, value := atomic.LoadInt64(&a.hits), a.counter.Value()
+	hits, value := a.hits.Rate(), a.counter.Rate()
```

Comment:
I've just realized that having two values with separate atomic-swaps here introduces a minor race condition where inconsistent data would be shown. But, probably not a big issue (or related to your change).